projects/Mallard/src/helpers/weather.ts

```diff
@@ -111,27 +111,39 @@ const ONE_HOUR = MS_IN_A_SECOND * SECS_IN_A_MINUTE * MINS_IN_AN_HOUR
 /**
  * Resolve weather location information and forecasts. Refetch weather
  * after that if app is foregrounded and data is more than an hour old.
- * (We can't use `setTimeout` because of https://github.com/facebook/react-native/issues/12981)
+ * (We can't use `setTimeout` because of
+ * https://github.com/facebook/react-native/issues/12981)
+ *
+ * `refreshWeather` forces an immediate refresh, can be from user request or
+ * when the location permission change and we can now fetch more
+ * precise weather.
  */
-export const resolveWeather = (() => {
+const { resolveWeather, refreshWeather } = (() => {
     // If `undefined`, weather has never been shown.
     let weather: Promise<Weather> | undefined
 
+    // We update the cache only if no other update happened in-between.
+    const update = async (client: ApolloClient<object>, fallback?: Weather) => {
+        if (weather == null) return
+        const newWeather = (weather = getWeather(fallback))
+        const value = await weather
+        if (weather == newWeather)
```

Comment:
These few lines aren't the easiest to discern but I see what's happening now. Perhaps a comments?

```suggestion
        // `weather` may have changed if this function was called again while `await weather` was resolving
        if (weather == newWeather)
```